flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,361 @@
+                            (slotSharingGroup, statistics) -> {
+                                int slotNum = statistics.getMax();
+                                paralleledExecutionSlotSharingGroupsMap.put(
+                                        slotSharingGroup,
+                                        createExecutionSlotSharingGroups(
+                                                slotSharingGroup, slotNum));
+                            });
+        }
+
+        private List<ExecutionSlotSharingGroup> createExecutionSlotSharingGroups(
+                SlotSharingGroup slotSharingGroup, int slotNum) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    new ArrayList<>(slotNum);
+            for (int i = 0; i < slotNum; i++) {
+                final ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        new ExecutionSlotSharingGroup(slotSharingGroup);
+                executionSlotSharingGroups.add(i, executionSlotSharingGroup);
+                LOG.debug(
+                        "Create {}th executionSlotSharingGroup {}.", i, executionSlotSharingGroup);
+            }
+            return executionSlotSharingGroups;
+        }
+
+        private void allocateCoLocatedVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices,
+                int index) {
+            final int slotNum =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup).size();
+            final int jobVertexParallel = executionVertices.size();
+            if (slotNum == jobVertexParallel) {
+                // For CoLocations with max parallelism of the slot sharing group.
+                allocateCoLocatedMaxParallelVertices(slotSharingGroup, executionVertices, index);
+            } else {
+                // For CoLocations with non-max parallelism of the slot sharing group.
+                allocateCoLocatedNonMaxParallelVertices(slotSharingGroup, executionVertices);
+            }
+        }
+
+        private void allocateCoLocatedMaxParallelVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices,
+                int index) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            for (SchedulingExecutionVertex executionVertex : executionVertices) {
+                final CoLocationConstraint coLocationConstraint =
+                        getCoLocationConstraint(executionVertex);
+                ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        constraintToExecutionSlotSharingGroupMap.get(coLocationConstraint);
+                if (Objects.isNull(executionSlotSharingGroup)) {
+                    executionSlotSharingGroup = executionSlotSharingGroups.get(index);
```

Comment:
Could we update `constraintToExecutionSlotSharingGroupMap` here directly?

It's null, we can update directly. Currently, `updateConstraintToExecutionSlotSharingGroupMap` call a series of same logic with here.

`allocateCoLocatedNonMaxParallelVertices` is same with here.